<?xml version="1.0"?>
<robot name="puzzlebot">

  <link name="base_link">
    <inertial>
      <mass value="0.0001"/> <!-- Negligible mass for dummy link -->
      <origin xyz="0 0 0" />
      <inertia ixx="0.0" ixy="0.0" ixz="0.0" iyy="0.0" iyz="0.0" izz="0.0"/>
    </inertial>
  </link>

  <!-- Chassis -->
  <link name="chassis">
  <!-- <pose relative_to='__model__'>0 0 0.05 0 0 0</pose> -->
    <inertial>
      <mass value="0.9"/> <!-- Chassis mass in kg -->
      <origin rpy="1.5708 0 1.5708" xyz="0 0 0.05" /> <!-- Centered in chassis, 4.3 cm height / 2 -->
      <inertia ixx="0.001" ixy="0.0" ixz="0.0" iyy="0.002" iyz="0.0" izz="0.002"/>
    </inertial>
    <visual>
      <origin rpy="1.5708 0 1.5708" xyz="0 0 0.05" />
      <geometry>
        <mesh filename="package://puzzlebot_challenge/models/MCR2_1000_13_Chassis.stl"/>
      </geometry>
      <material name="white">
        <color rgba="1.0 1.0 1.0 1.0"/>
      </material>
    </visual>
    <collision>
      <origin rpy="1.5708 0 1.5708" xyz="0 0 0.05" />
      <geometry>
        <mesh filename="package://puzzlebot_challenge/models/MCR2_1000_13_Chassis.stl"/>
      </geometry>
    </collision>
  </link>

  <!-- Left Wheel -->
  <link name="left_wheel">
    <inertial>
      <mass value="0.0471238898038469"/> <!-- Left wheel mass in kg -->
      <origin rpy="0 1.5708 -1.5708" xyz="0.0 0.0 0.0" />
      <inertia ixx="0.00001" ixy="0.0" ixz="0.0" iyy="0.00002" iyz="0.0" izz="0.00001"/>
    </inertial>
    <visual>
      <origin rpy="0 1.5708 -1.5708" xyz="0.0 0.0 0.0"/>
      <geometry>
        <mesh filename="package://puzzlebot_challenge/models/MCR2_1000_1_1_Wheel_Coupler_2.stl"/>
      </geometry>
      <material name="black">
        <color rgba="0 0 0 1.0"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 1.5708 -1.5708" xyz="0.0 0.0 0.0" />
      <geometry>
        <mesh filename="package://puzzlebot_challenge/models/MCR2_1000_1_1_Wheel_Coupler_2.stl"/>
      </geometry>
    </collision>
  </link>

  <!-- Right Wheel -->
  <link name="right_wheel">
    <inertial>
      <mass value="0.0471238898038469"/> <!-- Right wheel mass in kg -->
      <origin rpy="0 -1.5708 -1.5708" xyz="0.0 0.0 0.0" />
      <inertia ixx="0.00001" ixy="0.0" ixz="0.0" iyy="0.00002" iyz="0.0" izz="0.00001"/>
    </inertial>
    <visual>
      <origin rpy="0 -1.5708 -1.5708" xyz="0.0 0.0 0.0"/>
      <geometry>
        <mesh filename="package://puzzlebot_challenge/models/MCR2_1000_1_1_Wheel_Coupler_2.stl"/>
      </geometry>
      <material name="black">
        <color rgba="0 0 0 1.0"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 -1.5708 -1.5708" xyz="0.0 0.0 0.0" />
      <geometry>
        <mesh filename="package://puzzlebot_challenge/models/MCR2_1000_1_1_Wheel_Coupler_2.stl"/>
      </geometry>
    </collision>
  </link>


  <!-- RPLidar Link -->
  <link name="rplidar_link">
    <!-- <inertial>
      <mass value="0.1"/>
      <origin xyz="0 0 0.02" rpy="0 0 0"/>
      <inertia ixx="0.0001" ixy="0.0" ixz="0.0" iyy="0.0001" iyz="0.0" izz="0.0001"/>
    </inertial> -->
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <cylinder radius="0.035" length="0.027"/> <!-- Adjusted size for RPLidar -->
      </geometry>
      <material name="black">
        <color rgba="0.0 0.0 0.0 1.0"/>
      </material>
    </visual>
  </link>



  <!-- Joints -->
  <joint name="chassis_joint" type="fixed">
    <parent link="base_link"/>
    <child link="chassis"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
  </joint>
  
  <joint name="wheel_left_joint" type="continuous">
    <parent link="chassis"/>
    <child link="left_wheel"/>
    <origin xyz="0.05 0.085 0.05" rpy="0 0 0"/> <!-- Left wheel position -->
    <axis xyz="0 1 0"/>
  </joint>

  <joint name="wheel_right_joint" type="continuous">
    <parent link="chassis"/>
    <child link="right_wheel"/>
    <origin xyz="0.05 -0.085 0.05" rpy="0 0 0"/> <!-- Right wheel position -->
    <axis xyz="0 1 0"/>
  </joint>  

  <joint name="rplidar_joint" type="fixed">
    <parent link="chassis"/>
    <child link="rplidar_link"/>
    <origin xyz="0.05 0 0.16" rpy="0 0 0"/> <!-- Position RPLidar above chassis -->
  </joint>

  <gazebo>
    <plugin name="gz::sim::systems::DiffDrive" filename="ignition-gazebo-diff-drive-system">
      <left_joint>wheel_left_joint</left_joint>
      <right_joint>wheel_right_joint</right_joint>
      <wheel_separation>0.19</wheel_separation> <!-- Distance between wheels -->
      <wheel_radius>0.05</wheel_radius>          <!-- Wheel radius -->
      <topic>/cmd_vel</topic>                    <!-- Ignition topic for velocity commands -->
    </plugin>
  </gazebo>

  <gazebo reference="rplidar_link">
	<sensor type="gpu_lidar" name="rplidar">
		<pose>0 0 0 0 0 0</pose>
		<visualize>true</visualize>
		<update_rate>10</update_rate>
		<ray>
			<scan>
				<horizontal>
					<samples>720</samples>
					<resolution>1</resolution>
					<min_angle>-3.14159</min_angle>
					<max_angle>3.14159</max_angle>
				</horizontal>
			</scan>
			<range>
				<min>0.15</min>
				<max>6.0</max>
				<resolution>0.01</resolution>
			</range>
			<noise>
				<type>gaussian</type>
				<mean>0.0</mean>
				<stddev>0.01</stddev>
			</noise>
		</ray>
		<plugin name="ignition::gazebo::systems::Sensors" filename="libignition-gazebo-sensors-system.so">
			<topicName>/scan</topicName>
			<frameName>rplidar_link</frameName>
		</plugin>
	</sensor>
</gazebo>


</robot>